### PR TITLE
Implemented multiple new command parameters for CI/CD functionalities for Azure Webapps 

### PR DESCRIPTION
**Pull request by kuvinodms**
Implemented following new command parameters for CI/CD functionalities for Azure Webapps:

cd-account-create
app-working-directory
private-repo-username
private-repo-password
cd-project-url
test
swap-slot

Along with these, added support for PHP, Python and NodeJS Webapps.

### CODE REVIEW

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 1]
```diff
@@ -60,17 +61,21 @@ def set_azure_web_info(self, resource_group_name, website_name, credentials,
         self._azure_info.tenant_id = tenant_id
         self._azure_info.webapp_location = webapp_location
 
-    def set_repository_info(self, repo_url, branch, git_token):
+    def set_repository_info(self, repo_url, branch, git_token, private_repo_username, private_repo_password):
         """
         Call this method before attempting to setup continuous delivery to setup the source control settings
         :param repo_url:
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: Why is this comment there without any description

File: continuous_delivery/models/authorization_info_parameters.py [hunk 2]
```diff
@@ -12,8 +12,12 @@ class AuthorizationInfoParameters(Model):
     _attribute_map = {
         'authorization': {'key': 'Authorization', 'type': 'str'},
         'access_token': {'key': 'AccessToken', 'type': 'str'},
+        'username': {'key': 'username', 'type': 'str'},
+        'password': {'key': 'password', 'type': 'str'}
     }
 
-    def __init__(self, authorization=None, access_token=None):
+    def __init__(self, authorization=None, access_token=None, username=None, password=None):
         self.authorization = authorization
```
Comment (sachinma) [inline continuous_delivery/models/authorization_info_parameters.py:20]: What is authorization

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 3]
```diff
@@ -206,9 +242,14 @@ def _get_source_repository(self, uri, token, branch, cred):
         else:
             match = re.match(r'[htps]+\:\/\/github\.com\/(.+)', uri, re.IGNORECASE)
             if match:
-                type = 'Github'
-                identifier = match.group(1)
-                auth_info = AuthorizationInfo('PersonalAccessToken', AuthorizationInfoParameters(None, token))
+                if username is None and password is None:
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: Should you rather check that token is present

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 4]
```diff
@@ -187,7 +223,7 @@ def _get_build_configuration(self, app_type_details, working_directory):
             raise RuntimeError("The app_type %s was not understood. Accepted values: %s." % (app_type, accepted_app_types))
         return build_configuration
 
-    def _get_source_repository(self, uri, token, branch, cred):
+    def _get_source_repository(self, uri, token, branch, cred, username, password):
         # Determine the type of repository (TfsGit, github, tfvc, externalGit)
         # Find the identifier and set the properties; default to externalGit
         type = 'ExternalGit'
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: why should the default type be ExternalGit

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 5]
```diff
@@ -206,9 +242,14 @@ def _get_source_repository(self, uri, token, branch, cred):
...
-                identifier = match.group(1)
-                auth_info = AuthorizationInfo('PersonalAccessToken', AuthorizationInfoParameters(None, token))
+                if username is None and password is None:                    
+                    type = 'Github'
+                    identifier = match.group(1).replace(".git", "")
+                    auth_info = AuthorizationInfo('PersonalAccessToken', AuthorizationInfoParameters(None, token))
+                else:                    
+                    type = 'Git'
+                    identifier = uri
+                    auth_info = AuthorizationInfo('UsernamePassword', AuthorizationInfoParameters(None, None, username, password))
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: How are you passing username and password when it is not a github url

**Review by sachinma — CHANGES_REQUESTED** [review]

File: continuous_delivery/models/authorization_info_parameters.py (hunk 2 above)
Comment (kuvinodms) [reply continuous_delivery/models/authorization_info_parameters.py:20]: 'Bearer ' + vsts_app_auth_token

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 6]
```diff
@@ -206,9 +244,14 @@ def _get_source_repository(self, uri, token, branch, cred):
...
             match = re.match(r'[htps]+\:\/\/github\.com\/(.+)', uri, re.IGNORECASE)
             if match:
-                type = 'Github'
-                identifier = match.group(1)
-                auth_info = AuthorizationInfo('PersonalAccessToken', AuthorizationInfoParameters(None, token))
+                if token is not None:                    
+                    type = 'Github'
+                    identifier = match.group(1).replace(".git", "")
+                    auth_info = AuthorizationInfo('PersonalAccessToken', AuthorizationInfoParameters(None, token))
+                elif username is not None and password is not None:
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: should you just remove the elif part

File: tests/test_continuous_delivery_manager.py [hunk 7]
```diff
@@ -53,11 +53,11 @@ def test_set_azure_web_info(self):
 
     def test_set_repository_info(self):
         cdman = ContinuousDeliveryManager(None)
-        cdman.set_repository_info('repoUrl1', 'master1', 'token1')
+        cdman.set_repository_info('repoUrl1', 'master1', 'token1', None, None)
```
Comment (sachinma) [inline tests/test_continuous_delivery_manager.py]: Test with non-null username and password

File: tests/test_continuous_delivery_manager.py [hunk 8]
```diff
@@ -173,12 +177,13 @@ def _get_provisioning_config(self, status, status_message):
             CiResult(status, status_message))
         return ProvisioningConfiguration('abcd', None, None, ci_config)
 
-def create_cd_app_type_details_map(cd_app_type, nodejs_task_runner, python_framework, python_version):
+def create_cd_app_type_details_map(cd_app_type, nodejs_task_runner, python_framework, python_version, app_working_dir):
```
Comment (sachinma) [inline tests/test_continuous_delivery_manager.py:212]: where are the tests for slot swap, test or validation

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 9]
```diff
@@ -80,23 +85,28 @@ def remove_continuous_delivery(self):
         # TODO: this would be called by appservice web source-control delete
         return
 
-    def setup_continuous_delivery(self, azure_deployment_slot, app_type_details, vsts_account_name, create_account,
-                                  vsts_app_auth_token):
+    def setup_continuous_delivery(self, azure_deployment_slot, app_type_details, cd_project_url, create_account,
```
Comment (sachinma) [inline vsts_cd_manager/continuous_delivery_manager.py]: Can you rename azure_deployment_slot to swap_with_slot or something to indicate that this is not the slot on which the cd is configured

**Review by sachinma — APPROVED** [review]

**Pull request merged**